### ISSUE TITLE
[FEATURE] Add Hibernate EmbeddingStore

### ISSUE THREAD

**Issue by humcqc** [langchain4j/langchain4j#1037]
**Is your feature request related to a problem? Please describe.**

Add Hibernate EmbeddingStore.
In the context your application is already using hibernate with multitenancy, you would like all the tables to be handled in the same way.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

If possible take into account, authentication, multitenancy
Perhaps should be done in quarkus-langchain4J

**Comment by langchain4j** [comment]
@humcqc could you please provide a bit more details? How would that look like? Thanks

**Comment by humcqc** [comment]
Hi @langchain4j , 
I currently have a quarkus app using  OIDC  and hibernate discriminator multitenancy.
For my AI service inside this app I'm currently using PgVectorEmbedding Store (#851 ).

I would like all this to be coherent and when a user is chatting with my AI Assistant component the content retrieved from the store should be filtered by the tenant and by the data the user is allowed to see.

In my case this is done using hibernate. I can duplicate using just JDBC stuff , but I would like to find a better and integrated option.

For the moment I don't have solution, I was thinking to replace the create table by an hIbernate entity, and using hibernate queries instead of statement directly in the store. Then iterate to see if i can find a good solution.

**Comment by langchain4j** [comment]
Hi @humcqc, did you consider storing user/memory ID in the `Metadata` of a `TextSegment` on the ingestion phase and then using dynamic filter on the retrieval phase?

```java
interface AiService {

    String chat(@MemoryId String memoryId, @UserMessage String userMessage);
}

ContentRetriever contentRetriever = EmbeddingStoreContentRetriever.builder()
                .embeddingStore(pgVectorEmbeddingStore)
                .embeddingModel(embeddingModel)
                .dynamicFilter(query -> metadataKey("userId").isEqualTo(query.metadata().chatMemoryId().toString()))
                .build();
```

**Comment by humcqc** [comment]
Hi @langchain4j , 

My use case concerns condominiums.  A tenant is an organization that handle multiple condominiums. 
When a user living in a condominium is asking question to the AI assistant, I just want the AI to use the information from his condominium.

The user is authenticated with OIDC.
From OIDC i want hibernate to filter by tenant with the tenant resolver and use a metadata to filter by condominium.

I will not use MemoryId.

I will filter at some point some info with user roles, but I will use a metadata for this.

Regards

**Comment by langchain4j** [comment]
@humcqc are you referring to [hibernate-vector](https://central.sonatype.com/artifact/org.hibernate.orm/hibernate-vector) that you want to use?

**Comment by humcqc** [comment]
yes, something like https://www.youtube.com/watch?v=0Z84qAEzoZg

**Comment by langchain4j** [comment]
@humcqc I see, would you like to do a prototype of such `HibernateEmbeddingStore`?

**Comment by humcqc** [comment]
yes, I will try in my quarkus app , and see if i can move it down to langchain.
When is plan the next release? I need first to finish the PgVectorEmbeddingStore integration in quarkus-langchain

**Comment by langchain4j** [comment]
@humcqc that would be awesome, thank you!
Next release is hopefully next Tuesday or Webnesday

**Comment by zambrinf** [comment]
@humcqc any updates? Im thinking of working on it

**Comment by humcqc** [comment]
Hi @zambrinf , I didn't start yet. no pb If you want to do it.

**Comment by zambrinf** [comment]
@humcqc were you thinking of HibernateEmbeddingStore being like a proxy? How would it comply to the langchain4j EmbeddingStore interface in case of using multi tenancy?

**Comment by humcqc** [comment]
@zambrinf I was thinking implementing a langchain4j EmbeddingStore taking as input an User Application Entity complying with EmbeddingStore needed columns. And then the  HibernateEmbeddingStore will do the glue between EmbeddingStore and the entity.
